staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -18,146 +18,115 @@ import (
-type OpenAPISchema struct {
-	Title                string                    `yaml:"title,omitempty"`
-	Description          string                    `yaml:"description,omitempty"`
-	Type                 string                    `yaml:"type,omitempty"`
-	TypeParam            string                    `yaml:"type_param,omitempty"`
-	TypeRef              string                    `yaml:"$ref,omitempty"`
-	DefaultValue         interface{}               `yaml:"default,omitempty"`
-	Enum                 []interface{}             `yaml:"enum,omitempty"`
-	Format               string                    `yaml:"format,omitempty"`
-	Items                *OpenAPISchema            `yaml:"items,omitempty"`
-	Metadata             map[string]string         `yaml:"metadata,omitempty"`
-	Required             []string                  `yaml:"required,omitempty"`
-	Properties           map[string]*OpenAPISchema `yaml:"properties,omitempty"`
-	AdditionalProperties *OpenAPISchema            `yaml:"additionalProperties,omitempty"`
-}
 
-// DeclTypes constructs a top-down set of DeclType instances whose name is derived from the root
+// SchemaDeclTypes constructs a top-down set of DeclType instances whose name is derived from the root
 // type name provided on the call, if not set to a custom type.
-func (s *OpenAPISchema) DeclTypes(maybeRootType string) (*DeclType, map[string]*DeclType) {
-	root := s.DeclType().MaybeAssignTypeName(maybeRootType)
+func SchemaDeclTypes(s *schema.Structural, maybeRootType string) (*DeclType, map[string]*DeclType) {
+	root := SchemaDeclType(s).MaybeAssignTypeName(maybeRootType)
 	types := FieldTypeMap(maybeRootType, root)
 	return root, types
 }
 
-// DeclType returns the CEL Policy Templates type name associated with the schema element.
-func (s *OpenAPISchema) DeclType() *DeclType {
-	if s.TypeParam != "" {
-		return NewTypeParam(s.TypeParam)
-	}
+// SchemaDeclType returns the CEL Policy Templates type name associated with the schema element.
+func SchemaDeclType(s *schema.Structural) *DeclType {
 	declType, found := openAPISchemaTypes[s.Type]
 	if !found {
-		return NewObjectTypeRef("*error*")
+		return nil
+	}
+	if s.XIntOrString {
+		return NewObjectType("intOrString", map[string]*DeclField{
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
+	}
+	if s.XEmbeddedResource {
+		return NewObjectType("embedded", map[string]*DeclField{
+			"kind":       {Name: "kind", Type: StringType},
+			"apiVersion": {Name: "apiVersion", Type: StringType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name":         {Name: "name", Type: StringType},
+				"generateName": {Name: "generateName", Type: StringType, Required: false},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
+			"status": {Name: "status", Type: DynType, Required: false},
+		})
+	}
+	if s.XPreserveUnknownFields {
+		return DynType
```

Comment:
returning here omits type definitions we know about at this level (I can specify properties a, b, and c, AND set x-kubernetes-preserve-unknown-fields, and we'd want to stay informed about the type information for properties a, b, and c)

Reply:
Fixed in later commits.